server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+         */
+        SUCCESSFUL((byte) 1),
+        /**
+         * Tiering failed
+         */
+        FAILED((byte) 2);
+
+        private final byte value;
+
+        /**
+         * Constructs new state
+         *
+         * @param value state code
+         */
+        IndexTieringStatus(byte value) {
+            this.value = value;
+        }
+
+        /**
+         * Returns state code
+         *
+         * @return state code
+         */
+        public byte value() {
+            return value;
+        }
+
+        /**
+         * @return true if tiering is successful
+         */
+        public boolean successful() {
+            return this == SUCCESSFUL;
+        }
+
+        /**
+         * @return true if tiering  is failed
+         */
+        public boolean failed() {
+            return this == FAILED;
+        }
+
+        /**
+         * Returns state corresponding to state code
+         *
+         * @param value stat code
+         * @return state
+         */
+        public static IndexTieringStatus fromValue(byte value) {
+            switch (value) {
+                case 0:
+                    return INIT;
+                case 1:
+                    return SUCCESSFUL;
+                case 2:
+                    return FAILED;
+                default:
+                    throw new IllegalArgumentException("No tiering state for value [" + value + "]");
+            }
+        }
+    }
+
+    /**
+     * Shard Tiering state
+     *
+     * @opensearch.experimental
+     */
+    @ExperimentalApi
+    public enum ShardTieringState {
+        /**
+         * Initializing state
+         */
+        INIT((byte) 0),
+        /**
+         * Processing state
+         */
+        PROCESSING((byte) 1),
+        /**
+         * Tiering finished successfully
+         */
+        SUCCESSFUL((byte) 2),
+        /**
+         * Tiering failed
+         */
+        FAILED((byte) 3);
+
+        private final byte value;
+
+        /**
+         * Constructs new state
+         *
+         * @param value state code
+         */
+        ShardTieringState(byte value) {
+            this.value = value;
+        }
+
+        /**
+         * Returns state code
+         *
+         * @return state code
+         */
+        public byte value() {
+            return value;
+        }
+
+        /**
+         * Returns true if tiering completed (either successfully or with failure)
+         *
+         * @return true if tiering completed
+         */
+        public boolean completed() {
+            return this == SUCCESSFUL || this == FAILED;
+        }
+
+        /**
+         * Returns state corresponding to state code
+         *
+         * @param value stat code
+         * @return state
+         */
+        public static ShardTieringState fromValue(byte value) {
+            switch (value) {
+                case 0:
+                    return INIT;
+                case 1:
+                    return PROCESSING;
```

Comment:
Not a strong opinion -- should this mirror snapshot states and be `IN_PROGRESS`?
https://github.com/opensearch-project/OpenSearch/blob/087355f0ee676064ea409ed68090b33e568ea941/server/src/main/java/org/opensearch/snapshots/SnapshotState.java#L110-L113

Given this is all experimental as of now we don't need to decide right now.

Reply:
updated